### PR TITLE
fixing noncontig broadcast

### LINKED ISSUE #1741: Non-contiguous Tensor with Broadcast Bug
### 🐛 Describe the bug

This issue was originally provided by @ngimel.  This was her example using `refs`.  It is exposed via the Python Frontend.  This problem is not seen in TorchScript because the inputs are permuted to be contiguous in the integration.  So, a C++ test had to be written.

Original Test:

```
import torch
from functools import partial
from torch._prims.executor import make_traced
import torch._refs

def main():
    device="cuda"
    dtype=torch.half
    size=(4,32,16,112,112)
    size1=(32,1,112,1)
    a=torch.randn(size, device=device, dtype=dtype).transpose(-1,-2)
    b=torch.randn(size1, device=device, dtype=dtype).transpose(-1,-2)

    a+b
    def fn_add(a,b):
        return torch._refs.add(a,b)

    traced = make_traced(fn_add)
    for executor in ('aten','nvfuser'):
        fn = partial(traced, executor=executor)
        result = fn(a,b)
        result = fn(a,b)
        result = fn(a,b)
        expected = torch.add(a,b)
        print((result-expected).abs().max())

    torch.cuda.synchronize()

if __name__ == "__main__":
    main()
```

C++ Test:
```
TEST_F(NVFuserTest, FusionReproNoncontigBroadcast_CUDA) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());
  
  auto options = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
  at::Tensor t0 = at::randn({4, 32, 16, 112, 112}, options).transpose(-1, -2);
  at::Tensor t1 = at::randn({32, 1, 112, 1}, options).transpose(-1, -2);

  auto tv0 = TensorViewBuilder()
                 .ndims(5)
                 .contiguity({true, true, false, false, false}) // ttfff
                 .shape({4, 32, 16, 112, 112})
                 .dtype(DataType::Half)
                 .build();
  auto tv1 = TensorViewBuilder()
                 .ndims(4)
                 .contiguity({true, false, false, true}) // tfft
                 .shape({32, 1, 1, 112})
                 .dtype(DataType::Half)
                 .build();

  fusion->addInput(tv0);
  fusion->addInput(tv1);

  auto tv2 = add(tv0, tv1);

  fusion->addOutput(tv2);

  std::vector<IValue> aten_inputs({t0, t1});

  FusionExecutorCache executor_cache(std::move(fusion));
  auto cg_outputs = executor_cache.runFusionWithInputs(aten_inputs);

  auto t2 = t0 + t1;

  testValidate(
      executor_cache.fusion(),
      cg_outputs,
      {t0, t1},
      {t2},
      __LINE__,
      __FILE__);
}
```
Error:
```
C++ exception with description "stride == cur_contig_stride || (still_rightmost && stride == 1) || (!still_rightmost && stride % word_size == 0) INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/executor_utils.cpp":586, please report a bug to PyTorch. Vectorization of T1_g[ iS97{( ceilDiv(( ceilDiv(( 32 * ( 1 * ( 1 * 112 ) ) ), 4) ), blockDim.x) )}, iS96{4}, iS98{blockDim.x} ] with word size 4 not possible due to invalid stride. Domain: iS98{blockDim.x}, stride: 1
```

### Versions

TOT

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1741

Shamelessly stole Kevin's cpp repro :smiling_imp:

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 1]
```diff
@@ -582,12 +582,14 @@ void validateAlignedVectorizedFusionInputOutput(
...
+    const auto size = aten_tensor.sizes().at(i);
     // If this domain is contiguous, then not necessary to check the
     // stride. Otherwise, stride must be 1 if it's rightmost or
-    // divisible by word_size.
+    // divisible by word_size, unless its size == 1, which shouldn't
+    // affect the vectorization scheme.
     TORCH_INTERNAL_ASSERT(
         stride == cur_contig_stride || (still_rightmost && stride == 1) ||
-            (!still_rightmost && stride % word_size == 0),
+            (!still_rightmost && (stride % word_size == 0 || size == 1)),
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp]: I think it's safe to relax it even further as
```
        size == 1 || stride == cur_contig_stride || (still_rightmost && stride == 1) ||
            (!still_rightmost && stride % word_size == 0),
```

**Review by kevinstephano — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 2]
```diff
@@ -599,9 +601,12 @@ void validateAlignedVectorizedFusionInputOutput(
...
     // If the domain is size-1, the next domain is still considered
     // rightmost.
-    const auto size = aten_tensor.sizes().at(i);
     still_rightmost = still_rightmost && size == 1;
-    cur_contig_stride = stride * size;
+    // We do not update cur_contig_stride for size==1 dimensions,
+    // since we have specialized vectorization stride check for them
+    if (size != 1) {
+      cur_contig_stride = stride * size;
+    }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp:609]: nit
```suggestion
      cur_contig_stride *= size;
```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:609]: This is not equivalent, as `cur_contig_stride == stride` is not always true.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the fix.

@csarofeen removed automatic conversion of size-one IterDoamins to broadcast IterDomains (https://github.com/csarofeen/pytorch/pull/1738/files#diff-28d672948f070384f5d8c15ec70180f5ac1b7632d713af96dec1634fbdbb89ceL38-L49) in #1738 (which is not yet merged). Does it conflict with this PR? I guess `tv1` won't get broadcast domains?

**Comment by csarofeen** [comment]
The only point I had in that change was that the TensorView contructor shouldn't be responsible for changing the iter domain. We should also always be using TensorViewBuilder which will convert const size==1 dims.

**Review by csarofeen — APPROVED** [review]
LGTM

**Comment by jjsjann123** [comment]
Sounds like we are good with this changes. Local tests also passed. I'm merging this one

**Pull request merged**